scala/scala.bzl

```diff
@@ -494,6 +511,57 @@ def _scala_test_impl(ctx):
     _write_test_launcher(ctx, rjars)
     return _scala_binary_common(ctx, cjars, rjars)
 
+def _prep_grep_pattern_from(patterns):
+    combined_pattern = ""
+    for pattern in patterns:
+      combined_pattern += " -e {}\.class".format(pattern)
+    return combined_pattern
+
+def _discover_classes(ctx, patterns, archive):
+    discovered_classes = ctx.new_file(ctx.label.name + "_discovered_classes.txt")
+    ctx.action(
+      #We need _jdk to even run _jar. Depending on _jar is not enough with sandbox
+      inputs= [archive] + ctx.files._jar + ctx.files._jdk,
```

Comment:
TL;DR: Definitely prefer ctx.executable._jar here, especially since you are using it as an executable.

So, I did some major experimentation with this. _jar being marked as executable should be sufficient to automatically pull in its runfiles regardless of whether you access via executable or files. (ctx.action WILL NOT find runfiles on all inputs, just those marked with executable=True. Note that you can still find runfiles on any label, executable or not, via ctx.attr.....data_runfiles|default_runfiles.files. 

That said, I think it is safest to access via ctx.executable._jar so it is clear that it is being used as an executable and that there is an error if you accidentally remove executable=True.